### PR TITLE
Get the etcd data path from kubeadm config or etcd pod manifest on kubeadm reset

### LINKED ISSUE #1169: kubeadm reset doesn't clean up custom etcd datadir
## Versions
**kubeadm version** (use `kubeadm version`):
[root@myk8s ~]# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:14:39Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

## What happened?
Tried to init-reset-init which failed with the following preflight error:
```
[preflight] Some fatal errors occurred:
	[ERROR DirAvailable--path-to-etcd]: /path/to/etcd is not empty
```

## What you expected to happen?
The kubeadm reset should check whether the local etcd is using a custom dataDir and clean that directory instead of the default.
https://github.com/kubernetes/kubernetes/blob/488ac6825bacc766590adc9907b3cd5b327813f8/cmd/kubeadm/app/cmd/reset.go#L155

## How to reproduce it (as minimally and precisely as possible)?
```
$ cat kubeadmconfig.yaml
apiVersion: kubeadm.k8s.io/v1alpha3
kind: ClusterConfiguration
etcd:
  local:
    dataDir: "/path/to/etcd"
$ kubeadm init --config kubeadmconfig.yaml
...
$ kubeadm reset
...
$ kubeadm init --config kubeadmconfig.yaml
...
[preflight] Some fatal errors occurred:
	[ERROR DirAvailable--path-to-etcd]: /path/to/etcd is not empty
...
```

### PR DESCRIPTION
**Pull request by yagonobre**
**What type of PR is this?**
 /kind bug

**What this PR does / why we need it**:
This clean up the correct etcd data path on kubeadm reset

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/1169

**Does this PR introduce a user-facing change?**:

```release-note
Kubeadm reset now clean up custom etcd data path
```
/sig cluster-lifecycle

### CODE REVIEW

**Comment by neolit123** [comment]
/kind bug
/assign @fabriziopandini 

@yagonobre 
thank you, could you please add a release note explaining the change in one sentence.

**Review by neolit123 — COMMENTED** [review]
@yagonobre 

thank you for this bug fix PR!
i have added some minor comments but in general - LGTM.

**Comment by yagonobre** [comment]
@neolit123 updated

**Review by fabriziopandini — COMMENTED** [review]
@yagonobre thanks for this PR!

My first impression reading this PR is that the etcd folder detection process should initially try to read the kubeadm config and only as a fallback parse the etcd manifest. This could have several advantages, especially if we consider all the variants of etcd we supports (external etcd Vs local etcd Vs stacked etcd)
What do you think about?

@timothysc @detiber opinions?

**Comment by yagonobre** [comment]
@fabriziopandini This will just apply when used local/stacked etcd, we don't clean up data of external etcd, and both use the same spec. But where I can get the kubeadm config? in the configmap?

**Comment by fabriziopandini** [comment]
@yagonobre 
Yes, kubeadm config is a config map under kube-system namespace and we are using it as a "source of truth" about kubeadm clusters. Utilities for working with this config map are in `util\config` (even if some refactoring is still ongoing).

However, as suggested in my previous  I would appreciate some other opinion about this and then decide if to iterate on this PR or if to leave this PR as it is (+ eventually open an issue for keeping track of the possible improvement with kubeadm config map)

**Comment by timothysc** [comment]
> My first impression reading this PR is that the etcd folder detection process should initially try to read the kubeadm config and only as a fallback parse the etcd manifest. This could have several advantages, especially if we consider all the variants of etcd we supports (external etcd Vs local etcd Vs stacked etcd)
What do you think about? 

agreed

/assign @timothysc

**Comment by yagonobre** [comment]
@fabriziopandini @timothysc updated!

**Comment by fabriziopandini** [comment]
@yagonobre it seems you have to run /hack/update-bazel.sh to fix your test

**Comment by yagonobre** [comment]
@fabriziopandini done

**Comment by neolit123** [comment]
@yagonobre please squash to commits to 1 or 2 (one for code, one for autogenerated bazel).
thanks.

**Comment by yagonobre** [comment]
@neolit123 done

**Comment by neolit123** [comment]
LGTM
waiting on CI and further comments, if any.
@kubernetes/sig-cluster-lifecycle-pr-reviews

File: cmd/kubeadm/app/cmd/reset.go [hunk 1]
```diff
@@ -175,6 +185,33 @@ func (r *Reset) Run(out io.Writer) error {
 	return nil
 }
 
+func getEtcdDataDir(manifestPath string, client clientset.Interface) (string, error) {
+	const etcdVolumeName = "etcd-data"
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/reset.go]: const for this really shouldn't be localized.

File: cmd/kubeadm/app/cmd/reset.go [hunk 2]
```diff
@@ -175,6 +185,33 @@ func (r *Reset) Run(out io.Writer) error {
...
+
+	cfg, err := configutil.FetchConfigFromFileOrCluster(client, os.Stdout, "reset", "", false)
+	if err == nil {
+		return cfg.Etcd.Local.DataDir, nil
+	}
+	glog.Warningf("[reset] Unable to fetch the kubeadm-config ConfigMap, using etcd pod spec as fallback: %v", err)
+
+	etcdPod, err := utilstaticpod.ReadStaticPodFromDisk(manifestPath)
+	if err != nil {
+		return "", err
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/reset.go]: it's weird to return empty strings vs. nil objects.

**Review by timothysc — COMMENTED** [review]
Thx for the patch! 

Comments for next time, and they are non-blocking 

/lgtm

**Pull request merged**